Add `mapResults` option.

izaakschroeder opened the pull request:
**Changes**
- Clean up script to publish only necessary files to npm.
- Add `mapResults` option (closes #2)
- Add `getFilename` option and removed use of `loaderUtils.interpolateName` (closes #1)
- Preserve file name of rendered path (closes #4)
- Add generic class types to allow accurate typing between config functions.
- Improved support for relative `href` attributes in crawled markup.
- Change `mapStatsToProps` signature to receive raw stats object, allowing for custom json conversion.
- Updated visited record mechanism to consider the paths `/foo/bar` and `/foo/bar/index.html` to be the same.

10xjs commented on src/PagesPlugin.js:
```diff
@@ -3,21 +3,22 @@ import cheerio from 'cheerio';
+type OutputResult = {
+  markup: string,
+  path: string,
+};
```
we should include the output filename here as well to allow renaming the file before it is emitted

10xjs commented on src/PagesPlugin.js:
```diff
@@ -3,21 +3,22 @@ import cheerio from 'cheerio';
 type Options = {
   name: string,
   mapStatsToProps: (stats: Object) => Object,
+  mapResults: (Array<OutputResult>) => Array<OutputResult>,
```
`transformResults` is a more correct name for this

10xjs commented on src/PagesPlugin.js:
```diff
@@ -26,11 +27,7 @@ type DoItOptions = {
-  props,
-  paths = ['/'],
-}: DoItOptions) => {
+const doIt = ({render, props, paths = ['/']}: DoItOptions) => {
```
As good a time as any to give this function a name

10xjs commented on src/PagesPlugin.js:
```diff
@@ -41,55 +39,83 @@ const doIt = ({
+  name: '[path][name].[ext]',
+  paths: ['/'],
+  mapStatsToProps: () => {
+    throw new TypeError();
```
say "NO" to undefined error messages

10xjs commented on src/PagesPlugin.js:
```diff
@@ -3,34 +3,32 @@ import cheerio from 'cheerio';
+  markup: string,
+  path: string,
+};
+
```
ideas:
```js
type RenderResult = {
  markup: string,
  redirect?: string,
  status?: number,
}

type OutputResult = {
  ...RenderResult,
  path: string,
  filename: string,
}
```

10xjs replied:
... or better yet, make it generic

```js
class PagesPlugin<T: {markup: string}>  {
  constructor({render: (path, stats) => T});
}
```

10xjs commented:
Idea:

```js
new PagesPlugin({
  dirname: '[path][name]/index.[ext]',
  filename: '[path][name].[ext]',
  ...
});
```

Either `dirname` or `filename` is used depending on whether or not the rendered path has an extension.

This would also solve https://github.com/metalabdesign/pages-webpack-plugin/issues/4.

10xjs committed:
Upgrade flow version

10xjs committed:
Clean up script to publish only necessary files to npm

10xjs committed:
Refactor

**Changes**
- Add `mapResults` option (closes #2)
- Add `getFilename` option and removed use of `loaderUtils.interpolateName` (closes #1)
- Preserve file name of rendered path (closes #4)
- Add generic class types to allow accurate typing between config functions.
- Improved support for relative `href` attributes in crawled markup.
- Change `mapStatsToProps` signature to receive raw stats object, allowing for custom json conversion.
- Updated visited record mechanism to consider the paths `/foo/bar` and `/foo/bar/index.html` to be the same.

10xjs reviewed (approved):
🚢
